bundles/org.openhab.binding.ondilo/src/main/java/org/openhab/binding/ondilo/internal/OndiloApiClient.java

```diff
@@ -53,41 +61,25 @@ public OndiloApiClient(OAuthClientService oAuthService, AccessTokenResponse acce
     }
 
     @Nullable
-    public synchronized String get(String endpoint) {
-        try {
-            refreshAccessTokenIfNeeded();
-            URL url = new URI(ONDILO_API_URL + endpoint).toURL();
-            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
-            conn.setRequestMethod("GET");
-            conn.setRequestProperty("Authorization", "Bearer " + bearer);
-            conn.setRequestProperty("Accept", "application/json");
-            conn.setRequestProperty("Accept-Charset", "utf-8");
-            conn.setRequestProperty("Accept-Encoding", "gzip, deflate");
-            conn.connect();
-            int responseCode = conn.getResponseCode();
-            if (responseCode == 200) {
-                try (InputStream is = conn.getInputStream(); Scanner scanner = new Scanner(is, "UTF-8")) {
-                    return scanner.useDelimiter("\\A").next();
-                }
-            } else {
-                logger.warn("Ondilo API request failed with code: {}", responseCode);
+    public synchronized <T> T request(String requestMethod, String endpoint, Type type) {
+        // Enforce API rate limit: at least 200ms between requests
+        long now = System.currentTimeMillis();
+        long wait = lastRequestTime + MIN_REQUEST_INTERVAL_MS - now;
+        if (wait > 0) {
+            try {
+                Thread.sleep(wait);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                logger.debug("API request pause interrupted: {}", e.getMessage());
+                return null;
             }
-        } catch (InterruptedIOException e) {
-            logger.debug("Ondilo API request interrupted: {}", e.getMessage());
-            Thread.currentThread().interrupt();
-        } catch (IOException | URISyntaxException e) {
-            logger.warn("Ondilo API request error", e);
         }
-        return null;
-    }
-
-    @Nullable
-    public synchronized String put(String endpoint) {
+        lastRequestTime = System.currentTimeMillis();
```

Comment:
Can you place this any closer to the call? Preferable directy after the request is send?
Maybe a bit of a race condition, but between setting `lastRequestTime`  and sending the actual call, some milliseconds take place. By design this leads to the situation that the binding records a longer time between calls then the remote api.

Reply:
fixed